dev-packages/ovsx-client/src/ovsx-api-filter.ts

```diff
@@ -31,16 +36,55 @@ export interface OVSXApiFilter {
      * @param extensionId the extension id.
      * @returns the data for the latest compatible extension version if available, else `undefined`.
      */
-    getLatestCompatibleExtension(extensions: VSXExtensionRaw[]): VSXExtensionRaw | undefined;
+    getLatestCompatibleExtension(extensions: VSXExtensionRaw[], targetPlatform?: VSXTargetPlatform): VSXExtensionRaw | undefined;
     getLatestCompatibleVersion(searchEntry: VSXSearchEntry): VSXAllVersions | undefined;
 }
 
 export class OVSXApiFilterImpl implements OVSXApiFilter {
 
     constructor(
+        public client: OVSXClient,
         public supportedApiVersion: string
     ) { }
 
+    async findLatestCompatibleExtension(query: VSXQueryOptions): Promise<VSXExtensionRaw | undefined> {
+        const targetPlatform = query.targetPlatform;
+        if (!targetPlatform) {
+            return this.queryLatestCompatibleExtension(query);
+        }
+        const latestWithTargetPlatform = await this.queryLatestCompatibleExtension(query);
+        let latestUniversal: VSXExtensionRaw | undefined;
+        if (targetPlatform !== 'universal' && targetPlatform !== 'web') {
+            // Additionally query the universal version, as there might be a newer one available
+            latestUniversal = await this.queryLatestCompatibleExtension({ ...query, targetPlatform: 'universal' });
+        }
+        if (latestWithTargetPlatform && latestUniversal) {
+            // Prefer the version with the target platform if it's greater or equal to the universal version
+            return this.versionGreaterThanOrEqualTo(latestWithTargetPlatform.version, latestUniversal.version) ? latestWithTargetPlatform : latestUniversal;
+        }
+        return latestWithTargetPlatform ?? latestUniversal;
+    }
+
+    protected async queryLatestCompatibleExtension(query: VSXQueryOptions): Promise<VSXExtensionRaw | undefined> {
+        let offset = 0;
+        const size = 100;
```

Comment:
It's the default of the open vsx api for the `/v2/-/query` endpoint. It's just a reasonable default on our side in case the default of the open-vsx API ever changes. I'll add a comment to the code in question.

Reply:
I think the new code has an issue. I can see that the progress bar never disappears
![Peek 2024-06-24 09-53](https://github.com/eclipse-theia/theia/assets/5889696/91257101-03f7-451a-88e9-2967698609b9)
Also some other parts of the ui don't update as well, e.g. the install Button does not go to the installed state, although it works. 
I think one issue is that if there are no results `offset += results.extensions.length;` is still 0 and `loop = results.totalSize >= offset;` will always be true

Reply:
Oops, logic is difficult, that was actually supposed to be a `>`. Fixed now.